### DISCUSSION TITLE
Calculating ground state of Fermi-Hubbard model, and entanglement entropy as a function of length

### DISCUSSION THREAD

**Discussion by sbisw002** [QuSpin/QuSpin#467] [Q&A]
Hi. New user here. 1)For the calculation of ground state should the argument in eigsh be 'SM' or 'SA'?
2) For half-filling, should mu be 0 or U/2?  We already set N_up and N_down, so what does mu do?
3) How can I plot the ground state expectation value of particles vs site indices (  <n_i> vs i )?
4) I tried to calculate and plot the entanglement entropy, S(l) vs l. Is the code ok?

##### define model parameters #####
L=12 # system size
J=1.0 # hopping
U=4 # interaction
mu=0.0 # chemical potential

# define boson basis with 3 states per site L bosons in the lattice
N_up = L//2 # number of fermions with spin up
N_down = L//2 # number of fermions with spin down

print(N_up)
print(N_down)

#basis = spinful_fermion_basis_1d(L,Nf=(N_up,N_down),a=1,kblock=0,sblock=1)
basis = spinful_fermion_basis_1d(L,Nf=(N_up,N_down),kblock=0,sblock=1)

#basis = spinful_fermion_basis_1d(L,Nf=(N_up,N_down))

#print(basis)

# define site-coupling lists
hop_right=[[J,i,(i+1)%L] for i in range(L)] #PBC
hop_left= [[-J,i,(i+1)%L] for i in range(L)] #PBC 

#hop_right=[[J,i,(i+1)] for i in range(L-1)] #OBC
#hop_left= [[J,i,(i+1)] for i in range(L-1)] #OBC 

pot=[[-mu,i] for i in range(L)] # -\mu \sum_j n_{j \sigma}
interact=[[U,i,i] for i in range(L)] # U/2 \sum_j n_{j,up} n_{j,down}
# define static and dynamic lists
static=[
        ['+-|',hop_left],  # up hops left
        ['-+|',hop_right], # up hops right
        ['|+-',hop_left],  # down hops left
        ['|-+',hop_right], # down hops right
        ['n|',pot],        # up on-site potention
        ['|n',pot],        # down on-site potention
        ['n|n',interact]   # up-down interaction
                                ]
dynamic=[]
# build Hamiltonian
no_checks = dict(check_pcon=True,check_symm=True,check_herm=True)
H=hamiltonian(static,dynamic,basis=basis,dtype=np.float64,**no_checks)

# compute eigensystems of H1
#E1,V1=H.eigh()
#psi0=V1[:,0] # pick any state as initial state
# calculate entanglement entropy

E_GS,V_GS=H.eigsh(k=1,which='SM',maxiter=1E10) # only GS
print(np.shape(V_GS))

psi0 = V_GS
EnE = np.zeros(L-1)
S_L = np.zeros(L-1)

for Leng in range(1,L):
    env = [i for i in range(Leng)]
    Sent=ent_entropy(psi0,basis,(env,env) )
    print(Sent)
    print("Just printed")
    EnE[Leng-1] = Sent['Sent_A']
    S_L[Leng-1] = L-Leng
    print(L-Leng)

print(Sent['Sent_A'])

plt.plot(S_L,EnE,'ro')
plt.ylabel('entanglement entropy')
plt.show()
plt.close()

**Comment by mgbukov** [comment]
> 1)For the calculation of ground state should the argument in eigsh be 'SM' or 'SA'?

Check with scipy under eigsh -- whatever arguments are listed there you should be able to use in quspin. 

> 2) For half-filling, should mu be 0 or U/2? We already set N_up and N_down, so what does mu do?

If you fix Nup and N_down, then the mu-term will give yu a constant shift to the Hamiltonian. If you want to cosntruct Hamiltonian which obey particle-hole symmetry, etc., use the `z = n-1/2` operator string which effectively fixes mu. 

> 3) How can I plot the ground state expectation value of particles vs site indices ( <n_i> vs i )?

First, you have to construct the operators `n_i` (watch out -- they do not obey translation invariance, so no `kblock` symmetry is allowed); then you can use the `.expt_value(psi)` method of the `hamiltonian` class to compute the expectation. 

> 4) I tried to calculate and plot the entanglement entropy, S(l) vs l. Is the code ok?

I urge you to use the `basis.ent_entropy()` function instead -- see the documentation.

**Reply by sbisw002** [reply]
Thanks a lot for the prompt answer.
3. Can I have a pointer about how to construct operators?
4.
env1 = tuple(range(Leng) )
or env1=(0, 1, 2, 3)
Sent=basis.ent_entropy(V_GS,sub_sys_A=env1,density=True)['Sent_A']

gives the error message,

ValueError: sub_sys_A must be a tuple which contains the subsystems for the up spins in the first (left) part of the tuple and the down spins in the last (right) part of the tuple.

Can you please write a sample env1 or sub_sys_A?

Thanks again.
